src/Pact/Native/Trans/Types.hs

```diff
@@ -0,0 +1,130 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE ForeignFunctionInterface #-}
+
+-- |
+-- Module      :  Pact.Native.Trans.Types
+-- Copyright   :  (C) 2016 Stuart Popejoy
+-- License     :  BSD-style (see the file LICENSE)
+-- Maintainer  :  Stuart Popejoy <stuart@kadena.io>
+--
+-- Operators and math built-ins.
+--
+
+module Pact.Native.Trans.Types where
+
+import Data.Decimal (Decimal, normalizeDecimal)
+import Data.Int
+import Data.Word
+import Foreign.C.String (withCString, peekCString)
+import Foreign.C.Types
+import Foreign.ForeignPtr
+import Foreign.Marshal.Alloc (alloca, allocaBytes)
+import Foreign.Ptr
+import Foreign.Storable
+import System.IO.Unsafe (unsafePerformIO)
+
+type CPrecision = Int64
+type Sign = Int32
+type Exp = Int64
+type Limb = Word64
+
+data TransResult a
+  = TransNumber !a
+  | TransNaN
+  | TransInf
+  | TransNegInf
+  deriving (Functor, Foldable, Traversable)
+
+data MPFR = MP {
+  _precision :: {-# UNPACK #-} !CPrecision,
+  _sign      :: {-# UNPACK #-} !Sign,
+  _exponent  :: {-# UNPACK #-} !Exp,
+  _limbs     :: {-# UNPACK #-} !(ForeignPtr Limb)
+}
+
+instance Storable MPFR where
+    sizeOf _ = (32)
+    alignment _ = alignment (undefined :: Int64)
+    peek = error "MPFR.peek: Not needed and not applicable"
+    poke p (MP prec s e fp) = do
+      (\hsc_ptr -> pokeByteOff hsc_ptr 0) p prec
+      (\hsc_ptr -> pokeByteOff hsc_ptr 8) p s
+      (\hsc_ptr -> pokeByteOff hsc_ptr 16) p e
+      withForeignPtr fp $ \p1 -> (\hsc_ptr -> pokeByteOff hsc_ptr 24) p p1
+
+c'MPFR_RNDN :: CInt
+c'MPFR_RNDN = 0
+
+withFormattedNumber :: (Ptr CChar -> Ptr CChar -> IO a) -> IO a
+withFormattedNumber f =
+  allocaBytes 2048 $ \out ->
+    withCString "%.128R*f" $ \fmt -> f out fmt
+
+readResultNumber :: String -> TransResult Decimal
+readResultNumber "nan" = TransNaN
+readResultNumber "inf" = TransInf
+readResultNumber "-inf" = TransNegInf
+readResultNumber n = TransNumber (read (trimZeroes n))
+
+type Mpfr_t = Ptr MPFR
+
+foreign import ccall "mpfr_init"
```

Comment:
What do you think about marking these foreign imports as `unsafe`?

Reply:
We could, since it's always single threaded, but would the speed bump matter? Do we have a large enough number of such functions being used that there would be any reason to?

Reply:
Considering the overhead of ffi functions are now the basis for a lot of our math, I can see this adding up. However, I'm fine with doing an incremental perfening as-needed. Just want to confirm it's a conservative choice atm

Reply:
Yes, it was an intentionally conservative choice. They do invoke malloc and free underneath (for managing the limbs), but those should be thread-safe themselves.